src/gmail.js

```diff
@@ -1870,8 +1877,22 @@ var Gmail = function(localJQuery) {
 
         for (let email of email_data) {
             // cache email directly on IDs
-            c.emailIdCache[email.id] = email;
-            c.emailLegacyIdCache[email.legacy_email_id] = email;
+            if (c.emailIdCache[email.id] === undefined) {
+                //console.log("ADD c.emailIdCache[email.id]",c.emailIdCache[email.id],email);
+                c.emailIdCache[email.id] = email;
+            }
+            else if (data_source === "fd_request_payload" || data_source === "fd_embedded_json") {
+                //console.log("UPDATE c.emailIdCache[email.id]",c.emailIdCache[email.id],email,data_source);
+                c.emailIdCache[email.id] = email;
+            }
+            if (c.emailLegacyIdCache[email.legacy_email_id] === undefined) {
```

Comment:
Is updating the main cache and the legacy_cache independently really the best thing to do? Now we need to duplicate certain logic based on source.

Wouldn't it better to first add email to main cache, and then afterwards, based on the current contents of the main cache, add it to the legacy-cache? Something like this:

````javascript
}

const cachedEmail = c.emailidCache[email.id];
c.emailLegacyIdCache[cachedEmail.legacy_email_id] = cachedEmail;
````

Or would that cause some sort of errors which is being handled by the currently proposed implementation?

Reply:
You right, emailIdCache and emailLegacyIdCache must be load or update together, it's more secured. I changed that.